### PR TITLE
Incorporate initial code coverage support

### LINKED ISSUE #279: Add code coverage tools
We should add code coverage tools (probably [gcov](https://gcc.gnu.org/onlinedocs/gcc/Gcov.html)) to make sure we are testing appropriately.

### PR DESCRIPTION
**Pull request by joshessman-llnl**
This PR adds support for "interactive" code coverage analysis (recently updated in BLT) via LCOV with the `serac_coverage` target and for coverage report uploading to [codecov](https://codecov.io/gh/LLNL/serac/) (which is what Umpire uses).  It also adds a few sample error handling unit tests - these will get moved as part of #222.  These unit tests revealed an opportunity for inconsistency with coefficients that is partially addressed by #237 but is fixed here for now.

Instead of death tests (which don't play nice with the threading used by some of our TPLs, even in threadsafe mode), SLIC was configured to raise an exception on error which can be caught by GTest.

Resolves #279

### CODE REVIEW

File: src/physics/thermal_conduction.cpp [hunk 1]
```diff
@@ -51,7 +51,7 @@ void ThermalConduction::setTemperature(mfem::Coefficient& temp)
 void ThermalConduction::setTemperatureBCs(const std::set<int>&               temp_bdr,
                                           std::shared_ptr<mfem::Coefficient> temp_bdr_coef)
 {
-  bcs_.addEssential(temp_bdr, temp_bdr_coef, temperature_);
+  bcs_.addEssential(temp_bdr, temp_bdr_coef, temperature_, 0);
```
Comment (jamiebramwell) [inline src/physics/thermal_conduction.cpp]: I'm not sure we want to require this. It makes sense that a scalar coefficient on a scalar finite element state would not need a component index.

File: tests/serac_error_handling.cpp [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+#include "numerics/mesh_utils.hpp"
+#include "physics/utilities/boundary_condition.hpp"
+#include "physics/utilities/equation_solver.hpp"
+
+class SlicErrorException : public std::exception {
+};
+
+namespace serac {
+
+TEST(serac_error_handling, equationsolver_bad_lin_solver)
```
Comment (jamiebramwell) [inline tests/serac_error_handling.cpp:20]: Nice tests!

**Review by jamiebramwell — COMMENTED** [review]
Looks good! My only concern is requiring that scalar coefficients on scalar FE states require a component.

File: src/physics/thermal_conduction.cpp (hunk 1 above)
Comment (joshessman-llnl) [reply src/physics/thermal_conduction.cpp]: I think that's reasonable, it looks like the condition that should be changed then is here:
https://github.com/LLNL/serac/blob/9fc81ff51ce6a4b85a82e88e593f3bdc88eba771/src/physics/utilities/boundary_condition.cpp#L51-L63

I think the correct modification is to always assume scalar coef when the FEState's dimension is zero - does that sound right to you?

Comment (jamiebramwell) [reply src/physics/thermal_conduction.cpp]: Sounds good to me!

File: src/physics/utilities/boundary_condition.cpp [hunk 3]
```diff
@@ -59,7 +66,8 @@ void BoundaryCondition::project(FiniteElementState& state) const
       auto scalar_coef = std::get_if<std::shared_ptr<mfem::Coefficient>>(&coef_);
       SLIC_ASSERT_MSG(scalar_coef,
                       "Essential boundary condition contained a single component but had a non-scalar coefficient.");
-      state.gridFunc().ProjectCoefficient(**scalar_coef, dof_list, component_);
+      const int component = (vdim == 1) ? 0 : component_;
```
Comment (jamiebramwell) [inline src/physics/utilities/boundary_condition.cpp]: 👍

**Review by jamiebramwell — APPROVED** [review]

**Review by white238 — APPROVED** [review]

**Comment by joshessman-llnl** [comment]
LGTM

**Pull request merged**